zcash_primitives/src/transaction/components/amount.rs

```diff
@@ -0,0 +1,179 @@
+use std::iter::Sum;
+use std::ops::{Add, AddAssign, Sub, SubAssign};
+
+const COIN: i64 = 1_0000_0000;
+const MAX_MONEY: i64 = 21_000_000 * COIN;
+
+pub const DEFAULT_FEE: Amount = Amount(10000);
+
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub struct Amount(i64);
+
+impl Amount {
+    /// Returns a zero-valued Amount.
+    pub const fn zero() -> Self {
+        Amount(0)
+    }
+
+    /// Creates an Amount from an i64.
+    ///
+    /// Returns an error if the amount is out of range.
+    pub fn from_i64(amount: i64, allow_negative: bool) -> Result<Self, ()> {
```

Comment:
This is a minor nit, but I think that having an `allow_negative: bool` parameter isn't the best API design.  The reason is that when reading caller code, all that appears is
```rust
let amt = Amount::from_i64(val, true)?;
```
or similar, and the reader has to remember that the `true` means that negative values are **allowed**, rather than `true` meaning that the value is required to be positive (and therefore negative values are **disallowed**), which would be an alternate way to construct the API.

There are two alternatives that I think would improve readability in the long run:

1.  (my preference) Split this code into two constructors whose names (strawman proposal) indicate their behaviour:
```rust
pub fn from_i64_nonnegative(amount: i64) -> Result<Self, ()>;
pub fn from_i64_allow_negative(amount: i64) -> Result<Self, ()>;
```

2. Add an `enum AmountValidation { NonNegative, NegativeAllowed }` and replace the `bool` with an `AmountValidation`, so that caller code displays the validation criteria at the callsite:
```rust
let amt = Amount::from_i64(val, AmountValidation::NegativeAllowed)?;
```

Reply:
This is an excellent point. I also prefer the explicit constructors.